File: examples/02-modern-hpke/python/main.py
#!/usr/bin/env python3
"""
Applied PQC Lab - RFC 9180 HPKE (Hybrid Public Key Encryption) in Python
Suite: DHKEM(X25519, HKDF-SHA256) + HKDF-SHA256 + AES-256-GCM (Base Mode)
"""

import os
from cryptography.hazmat.primitives.asymmetric import x25519
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat


def hpke_labeled_extract(salt: bytes, label: bytes, ikm: bytes) -> bytes:
    labeled_ikm = b"HPKE-v1" + label + ikm
    hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=salt, info=b"")
    return hkdf.derive(labeled_ikm)


def hpke_labeled_expand(prk: bytes, label: bytes, info: bytes, length: int) -> bytes:
    labeled_info = length.to_bytes(2, "big") + b"HPKE-v1" + label + info
    hkdf = HKDF(algorithm=hashes.SHA256(), length=length, salt=None, info=labeled_info)
    # Using HKDF Expand directly via extract/expand
    from cryptography.hazmat.primitives.kdf.hkdf import HKDFExpand
    return HKDFExpand(algorithm=hashes.SHA256(), length=length, info=labeled_info).derive(prk)


def setup_base_sender(receiver_pubkey_bytes: bytes, info: bytes = b""):
    # 1. Generate ephemeral keypair
    sender_ephemeral_priv = x25519.X25519PrivateKey.generate()
    sender_ephemeral_pub = sender_ephemeral_priv.public_key()
    enc = sender_ephemeral_pub.public_bytes(Encoding.Raw, PublicFormat.Raw)

    # 2. Compute DH shared secret
    receiver_pub = x25519.X25519PublicKey.from_public_bytes(receiver_pubkey_bytes)
    dh_shared = sender_ephemeral_priv.exchange(receiver_pub)

    # 3. KEM Key Derivation (RFC 9180 DHKEM)
    kem_context = enc + receiver_pubkey_bytes
    shared_secret = hpke_labeled_extract(salt=b"", label=b"shared_secret", ikm=dh_shared)
    key_schedule_secret = hpke_labeled_extract(salt=shared_secret, label=b"secret", ikm=kem_context + info)

    key = hpke_labeled_expand(key_schedule_secret, label=b"key", info=b"", length=32)
    base_nonce = hpke_labeled_expand(key_schedule_secret, label=b"base_nonce", info=b"", length=12)

    return enc, key, base_nonce


def setup_base_receiver(enc_bytes: bytes, receiver_priv: x25519.X25519PrivateKey, info: bytes = b""):
    sender_ephemeral_pub = x25519.X25519PublicKey.from_public_bytes(enc_bytes)
    receiver_pub_bytes = receiver_priv.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)

    # 1. Compute DH shared secret
    dh_shared = receiver_priv.exchange(sender_ephemeral_pub)

    # 2. KEM Key Derivation
    kem_context = enc_bytes + receiver_pub_bytes
    shared_secret = hpke_labeled_extract(salt=b"", label=b"shared_secret", ikm=dh_shared)
    key_schedule_secret = hpke_labeled_extract(salt=shared_secret, label=b"secret", ikm=kem_context + info)

    key = hpke_labeled_expand(key_schedule_secret, label=b"key", info=b"", length=32)
    base_nonce = hpke_labeled_expand(key_schedule_secret, label=b"base_nonce", info=b"", length=12)

    return key, base_nonce


def main():
    print("======================================================")
    print(" [Applied PQC Lab] RFC 9180 HPKE Base Mode in Python")
    print(" Suite: DHKEM(X25519) + HKDF-SHA256 + AES-256-GCM")
    print("======================================================")

    # 1. Receiver generates static X25519 keypair
    print("[+] Step 1: Generating Receiver's static X25519 keypair...")
    receiver_priv = x25519.X25519PrivateKey.generate()
    receiver_pub_bytes = receiver_priv.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    print(f"    Receiver Public Key: {receiver_pub_bytes.hex()[:32]}... ({len(receiver_pub_bytes)} bytes)")

    # 2. Sender runs SetupBaseS
    app_info = b"Applied-PQC-Lab-HPKE-Context"
    aad = b"Authenticated-Session-Metadata-v1"
    plaintext = b"Confidential Payload secured by RFC 9180 HPKE Base Mode."

    print("[+] Step 2: Sender executing SetupBaseS(pkR)...")
    enc, sender_key, sender_nonce = setup_base_sender(receiver_pub_bytes, info=app_info)
    print(f"    Encapsulated Key (enc): {enc.hex()[:32]}... ({len(enc)} bytes)")

    # 3. Sender seals (encrypts) payload
    print("[+] Step 3: Sender sealing payload with AES-256-GCM...")
    aesgcm = AESGCM(sender_key)
    ciphertext = aesgcm.encrypt(sender_nonce, plaintext, associated_data=aad)
    print(f"    Ciphertext + Tag size: {len(ciphertext)} bytes")

    # 4. Receiver runs SetupBaseR
    print("[+] Step 4: Receiver executing SetupBaseR(enc, skR)...")
    receiver_key, receiver_nonce = setup_base_receiver(enc, receiver_priv, info=app_info)

    assert receiver_key == sender_key, "Derived AEAD key mismatch!"
    assert receiver_nonce == sender_nonce, "Derived base nonce mismatch!"
    print("    [PASS] Key Schedule secrets perfectly synchronized!")

    # 5. Receiver opens (decrypts & authenticates) ciphertext
    print("[+] Step 5: Receiver opening payload with AAD verification...")
    receiver_aesgcm = AESGCM(receiver_key)
    decrypted_msg = receiver_aesgcm.decrypt(receiver_nonce, ciphertext, associated_data=aad)

    assert decrypted_msg == plaintext, "Plaintext mismatch!"
    print("======================================================")
    print(f" [PASS] Decrypted message: \"{decrypted_msg.decode('utf-8')}\"")
    print(" [SUCCESS] RFC 9180 HPKE Base Mode verified in Python!")
    print("======================================================")


if __name__ == "__main__":
    main()

File: examples/02-modern-hpke/rust/src/main.rs
// =============================================================================
// Applied PQC Lab - RFC 9180 HPKE Base Mode in Rust (2021/2024 Edition)
// Suite: DHKEM(X25519) + HKDF-SHA256 + AES-256-GCM
// =============================================================================

use openssl::derive::Deriver;
use openssl::hash::MessageDigest;
use openssl::pkey::{Id, PKey};
use openssl::sign::Signer;
use openssl::symm::{decrypt_aead, encrypt_aead, Cipher};

fn hmac_sha256(key: &[u8], data: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let pkey = PKey::hmac(key)?;
    let mut signer = Signer::new(MessageDigest::sha256(), &pkey)?;
    signer.update(data)?;
    Ok(signer.sign_to_vec()?)
}

fn hkdf_extract(salt: &[u8], ikm: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let salt = if salt.is_empty() { &[0u8; 32] } else { salt };
    hmac_sha256(salt, ikm)
}

fn hkdf_expand(prk: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut info_with_counter = info.to_vec();
    info_with_counter.push(0x01);
    let okm = hmac_sha256(prk, &info_with_counter)?;
    Ok(okm[..len].to_vec())
}

fn hpke_labeled_extract(salt: &[u8], label: &[u8], ikm: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut labeled_ikm = b"HPKE-v1".to_vec();
    labeled_ikm.extend_from_slice(label);
    labeled_ikm.extend_from_slice(ikm);
    hkdf_extract(salt, &labeled_ikm)
}

fn hpke_labeled_expand(prk: &[u8], label: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut labeled_info = (len as u16).to_be_bytes().to_vec();
    labeled_info.extend_from_slice(b"HPKE-v1");
    labeled_info.extend_from_slice(label);
    labeled_info.extend_from_slice(info);
    hkdf_expand(prk, &labeled_info, len)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("======================================================");
    println!(" [Applied PQC Lab] RFC 9180 HPKE Base Mode in Rust");
    println!(" Suite: DHKEM(X25519) + HKDF-SHA256 + AES-256-GCM");
    println!("======================================================");

    // Step 1: Generate Receiver's static X25519 keypair
    println!("[+] Step 1: Generating Receiver's static X25519 keypair...");
    let receiver_pkey = PKey::generate_x25519()?;
    let receiver_pub_bytes = receiver_pkey.raw_public_key()?;
    println!("    Receiver Public Key size: {} bytes", receiver_pub_bytes.len());

    // Step 2: Sender executes SetupBaseS (Generate ephemeral key & DH)
    println!("[+] Step 2: Sender executing SetupBaseS(pkR)...");
    let sender_ephemeral = PKey::generate_x25519()?;
    let enc = sender_ephemeral.raw_public_key()?;

    let mut deriver = Deriver::new(&sender_ephemeral)?;
    deriver.set_peer(&receiver_pkey)?;
    let dh_shared = deriver.derive_to_vec()?;

    let mut kem_context = enc.clone();
    kem_context.extend_from_slice(&receiver_pub_bytes);

    let shared_secret = hpke_labeled_extract(&[], b"shared_secret", &dh_shared)?;
    let app_info = b"Applied-PQC-Lab-HPKE-Context";
    let mut secret_ikm = kem_context;
    secret_ikm.extend_from_slice(app_info);
    let key_schedule_secret = hpke_labeled_extract(&shared_secret, b"secret", &secret_ikm)?;

    let sender_key = hpke_labeled_expand(&key_schedule_secret, b"key", &[], 32)?;
    let sender_nonce = hpke_labeled_expand(&key_schedule_secret, b"base_nonce", &[], 12)?;
    println!("    Encapsulated Key (enc) size: {} bytes", enc.len());

    // Step 3: Sender seals payload with AES-256-GCM
    println!("[+] Step 3: Sender sealing payload with AES-256-GCM...");
    let plaintext = b"Confidential Payload secured by RFC 9180 HPKE Base Mode.";
    let aad = b"Authenticated-Session-Metadata-v1";
    let cipher = Cipher::aes_256_gcm();
    let mut tag = [0u8; 16];
    let ciphertext = encrypt_aead(cipher, &sender_key, Some(&sender_nonce), aad, plaintext, &mut tag)?;
    println!("    Ciphertext size: {} bytes, Tag: 16 bytes", ciphertext.len());

    // Step 4: Receiver executes SetupBaseR
    println!("[+] Step 4: Receiver executing SetupBaseR(enc, skR)...");
    let sender_pub_peer = PKey::public_key_from_raw_bytes(&enc, Id::X25519)?;
    let mut rec_deriver = Deriver::new(&receiver_pkey)?;
    rec_deriver.set_peer(&sender_pub_peer)?;
    let rec_dh_shared = rec_deriver.derive_to_vec()?;

    let mut rec_kem_context = enc;
    rec_kem_context.extend_from_slice(&receiver_pub_bytes);

    let rec_shared_secret = hpke_labeled_extract(&[], b"shared_secret", &rec_dh_shared)?;
    let mut rec_secret_ikm = rec_kem_context;
    rec_secret_ikm.extend_from_slice(app_info);
    let rec_key_schedule_secret = hpke_labeled_extract(&rec_shared_secret, b"secret", &rec_secret_ikm)?;

    let receiver_key = hpke_labeled_expand(&rec_key_schedule_secret, b"key", &[], 32)?;
    let receiver_nonce = hpke_labeled_expand(&rec_key_schedule_secret, b"base_nonce", &[], 12)?;

    assert_eq!(sender_key, receiver_key, "AEAD key mismatch!");
    assert_eq!(sender_nonce, receiver_nonce, "Base nonce mismatch!");
    println!("    [PASS] Receiver derived matching key and base nonce!");

    // Step 5: Receiver opens ciphertext
    println!("[+] Step 5: Receiver opening payload with AAD verification...");
    let decrypted_bytes = decrypt_aead(cipher, &receiver_key, Some(&receiver_nonce), aad, &ciphertext, &tag)?;
    let decrypted_msg = String::from_utf8(decrypted_bytes)?;

    println!("======================================================");
    println!(" [PASS] Decrypted message: \"{}\"", decrypted_msg);
    println!(" [SUCCESS] RFC 9180 HPKE Base Mode verified in Rust!");
    println!("======================================================");

    Ok(())
}
